lib/inventory/metadata.go

```diff
@@ -0,0 +1,314 @@
+		c.kubeClient = getKubeClient()
+	}
+}
+
+// getKubeClient returns a kubernetes client in case the agent is running on
+// kubernetes. It returns nil otherwise.
+func getKubeClient() kubernetes.Interface {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		log.Debugf("Failed to get kubernetes cluster config: %s", err)
+		return nil
+	}
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Debugf("Failed to create kubernetes client: %s", err)
+		return nil
+	}
+	return client
+}
+
+// fetchAgentMetadata fetches and calculates all agent metadata we are interested
+// in tracking.
+func fetchAgentMetadata(c *fetchConfig) proto.UpstreamInventoryAgentMetadata {
+	c.setDefaults()
+	return proto.UpstreamInventoryAgentMetadata{
+		Version:               c.fetchVersion(),
+		HostID:                c.fetchHostID(),
+		Services:              c.fetchServices(),
+		OS:                    c.fetchOS(),
+		OSVersion:             c.fetchOSVersion(),
+		HostArchitecture:      c.fetchHostArchitecture(),
+		GLibCVersion:          c.fetchGlibcVersion(),
+		InstallMethods:        c.fetchInstallMethods(),
+		ContainerRuntime:      c.fetchContainerRuntime(),
+		ContainerOrchestrator: c.fetchContainerOrchestrator(),
+		CloudEnvironment:      c.fetchCloudEnvironment(),
+	}
+}
+
+// fetchVersion returns the Teleport version present in the hello message.
+func (c *fetchConfig) fetchVersion() string {
+	return c.hello.Version
+}
+
+// fetchHostID returns the agent ID present in the hello message.
+func (c *fetchConfig) fetchHostID() string {
+	return c.hello.ServerID
+}
+
+// fetchOS returns the value of GOOS.
+func (c *fetchConfig) fetchOS() string {
+	return runtime.GOOS
+}
+
+// fetchServices computes the list of access protocols enabled at the agent from
+// the list of system roles present in the hello message.
+func (c *fetchConfig) fetchServices() []string {
+	var services []string
+	for _, svc := range c.hello.Services {
+		switch svc {
+		case types.RoleNode:
+			services = append(services, nodeService)
+		case types.RoleKube:
+			services = append(services, kubeService)
+		case types.RoleApp:
+			services = append(services, appService)
+		case types.RoleDatabase:
+			services = append(services, dbService)
+		case types.RoleWindowsDesktop:
+			services = append(services, windowsDesktopService)
+		}
+	}
+	return services
+}
+
+// fetchHostArchitecture computes the host architecture using the arch
+// command-line utility.
+func (c *fetchConfig) fetchHostArchitecture() string {
+	command := "arch"
+	out, err := c.exec(command)
+	if err != nil {
+		return ""
+	}
+
+	if !matchHostArchitecture.MatchString(out) {
+		return invalid(command, out)
+	}
+
+	return out
+}
+
+func (c *fetchConfig) fetchInstallMethods() []string {
+	// TODO(vitorenesduarte): fetch install methods
+	return []string{}
+}
+
+// fetchContainerRuntime returns "docker" if the file "/.dockerenv" exists.
+func (c *fetchConfig) fetchContainerRuntime() string {
```

Comment:
Why not use something like https://pkg.go.dev/github.com/genuinetools/bpfd@v0.0.1/proc#GetContainerRuntime to get more complete info?

Reply:
For this initial phase, the RFD proposes that we only try to track Docker. Happy to add this now if we feel that is valuable.

(`gopsutil` also detects a lot of runtimes: https://github.com/shirou/gopsutil/blob/v3.23.1/internal/common/common_linux.go#L130-L278)

Reply:
`genuinetools/bpfd` won't detect modern docker versions since it relies on `/proc/foo/cgroup` being populated, which is no longer the case. `gopsutil` looks like it should work, though.